### PR TITLE
[Instagram] Add new [:tag] extractor, refactor [:user]

### PR DESCRIPTION
**Pull request by jhwgh1968**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [X] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This adds an extractor for Instagram videos. It downloads all videos that match a given hashtag as a playlist. It is heavily based on the Instagram user extractor.

I did consider refactoring common methods instead of copy-pasting, but decided it was more important to avoid touching the other code. This way is more difficult to maintain, but I hope is easier to review and avoids the risk of regressions in the other extractors.

If you would prefer a refactor -- either now or in a separate PR -- I would be happy to do it.

### CODE REVIEW

**Comment by dstftw** [comment]
No copy pasting here.

File: youtube_dl/extractor/instagram.py [hunk 1]
```diff
@@ -227,6 +227,112 @@ def get_count(key, kind):
...
+                    'Downloading JSON page %d' % page_num, headers={
+                        'X-Requested-With': 'XMLHttpRequest',
+                        'X-Instagram-GIS': hashlib.md5(
+                            ('%s:%s' % (gis_tmpl, variables)).encode('utf-8')).hexdigest(),
+                    }, query={
+                        'query_hash': query_hash,
+                        'variables': variables,
+                    })
+                media = timeline_fn(query_json)
+                gis_tmpl = gis_tmpl
```
Comment (dstftw) [inline youtube_dl/extractor/instagram.py]: `_gis_tmpl` broken.

File: youtube_dl/extractor/instagram.py [hunk 2]
```diff
@@ -245,117 +351,65 @@ class InstagramUserIE(InfoExtractor):
...
+        assert tag
 
-        webpage = self._download_webpage(url, username)
+        webpage = self._download_webpage(url, tag)
 
         data = self._parse_json(
             self._search_regex(
                 r'sharedData\s*=\s*({.+?})\s*;\s*[<\n]', webpage, 'data'),
-            username)
+            tag)
```
Comment (dstftw) [inline youtube_dl/extractor/instagram.py]: DRY.

File: youtube_dl/extractor/instagram.py [hunk 3]
```diff
@@ -245,117 +351,65 @@ class InstagramUserIE(InfoExtractor):
...
         return self.playlist_result(
-            self._entries(data), username, username)
+            extract_graphql(self, data,
+                            _id=tag,
+                            query_hash='f92f56d47dc7a55b606908374b43a314',
+                            query_vars={'tag_name': tag},
+                            timeline_fn=lambda j:
+                                j['data']['hashtag']['edge_hashtag_to_media']
+                            ),
+            tag, tag)
```
Comment (dstftw) [inline youtube_dl/extractor/instagram.py]: DRY.

File: youtube_dl/extractor/instagram.py (hunk 1 above)
Comment (jhwgh1968) [reply youtube_dl/extractor/instagram.py]: I didn't understand it was a persistent cache at first. Fixed and commented.

File: youtube_dl/extractor/instagram.py (hunk 3 above)
Comment (jhwgh1968) [reply youtube_dl/extractor/instagram.py]: Where is the repetition? All of the call arguments are different between the two classes.

**Comment by jhwgh1968** [comment]
I have fixed two of the issues @dstftw, but I need clarification on the third one. Where is the repetition?

**Comment by jhwgh1968** [comment]
I have reworked the classes as you requested, @dstftw. I am pretty sure there is nothing left to repeat now.

**Comment by jhwgh1968** [comment]
Ping @dstftw, I think I have fixed everything. Is there anything else I need to do?

**Pull request merged**